docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -0,0 +1,210 @@
+
+Having a mechanism to sync an initial set of default resources after clusters are created makes clusters created with Cluster API functional and ready for workloads from the beginning, without requiring additional user intervention. 
+
+To achieve this, ClusterResourceSet CRD is introduced that will be responsible for applying a set resources defined by users to the matching clusters.
+
+### Goals
+
+- Provide a means to specify a set of resources to apply automatically to newly-created and existing Clusters. Resources initially will be applied only once; full sync functionality could be added later.
+- Support additions to the resource list by applying the new added resources to both new and existing matching clusters.
+- Support both json and yaml resources.
+
+### Non-Goals/Future Work
+
+- Replace or compete with the Cluster Addons subproject
+- Support deletion of resources from clusters. Deleting a resource from a ClusterResourceSet or deleting a ClusterResourceSet does not result in deletion of those resources from clusters.
+- Lifecycle management of the installed resources (such as CNI)
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As someone creating multiple clusters, I want all my clusters to have a CNI provider of my choosing installed automatically, so I don’t have to manually repeat the installation for each new cluster.
+
+#### Story 2
+
+As someone creating multiple clusters, I want all my clusters to have a StorageClass installed automatically, so I don't have to manually repeat the installation for each new cluster.
+
+#### Story 3
+
+As someone creating multiple clusters, I want to be able to provide different values for some fields in the resources for different clusters. For example, CNIs podCIDRs  may be required to be distinct for each cluster, hence some templating mechanism for variable substitution in the resources is needed.
+
+### Implementation Details/Notes/Constraints
+
+#### Data model changes
+
+None. We are planning to implement this feature without modifying any of the existing structure to minimize the footprint of ClusterResourceSet Controller. This enhancement will follow Kubernetes’s feature-gate structure and will be under the experimental package with its APIs, and enabled/disabled with a feature gate. 
+
+#### ClusterResourceSet Object Definition
+
+This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.
+
+The resources field is a list of secrets/configmaps in the same namespace. The clusterSelector field is a Kubernetes [label selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) that matches against labels on clusters (in any namespace).
+
+*Sample ClusterResourceSet YAML*
+
+```yaml
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: ClusterResourceSet
+metadata:
+ name: postcreate-conf
+ namespace: default
```

Comment:
Should we have a pre-defined namespace that we want all our ClusterResourceSets to be defined in? For example, `capi-resources-system`, or something similar

From a security standpoint having the ability to create a `ClusterResourceSet` in any namespace would give any users the ability to install something one every cluster.

Reply:
I'm not sure we should enter in this field.
Cluster admin should leverage on RBAC for defining their own preferred approach to this problem.

Reply:
Are we expecting that ClusterResourceSet are moved by clusterctl move?
If yes, we should check how the current implementation deals with experimental API types and with a possible use case where objects are spread across namespaces...

Reply:
I'm not sure we should enter in this field.
Cluster admin should leverage on RBAC for defining their own preferred approach to this problem.

Reply:
Also, are we expecting that ClusterResourceSet are moved by clusterctl move?
If yes, we should check how the current implementation deals with experimental API types and with a possible use case where ClusterResourceSet and clusters are spread across namespaces.
Also, the ownership chain of objects is relevant to move...

Reply:
> Cluster admin should leverage on RBAC for defining their own preferred approach to this problem.

RBAC is not capable of enforcing cross namespace boundaries nor is it able to enforce field level restrictions.

Reply:
Restricted cluster matching to the same namespace as the ClusterResourceSet so that the security concerns raised in the comments will be solved, and also `clusterctl move` will work.